Add dynamic page content preview

membralala opened the pull request:
### Short description
<!-- Describe this PR in one or two sentences. -->
This PR enhances the previous preview functionality in several ways.


### Proposed changes
<!-- Describe this PR in more detail. -->

- Dynamically render the preview as an overlay (similar to the page tree tutorial) 
- Add preview functionality to the page form.  

Note, that the preview in page tree will be generated based on the current saved state of the page in the database, but the preview in page form will be generated based on the edit state, even if it is not saved yet.

### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1056

membralala commented:
I'm not quite happy about the way the mirrored page content is fetched async in the page form. Maybe someone has suggestions on how to improve this?

timoludwig commented on integreat_cms/static/src/js/pages/page-preview.ts:
```diff
@@ -0,0 +1,82 @@
+        pageContent = tinymce.activeEditor.getContent();
+
+        // fetch mirrored page content based on the current dropdown value states
+        const mirroredPageRegion = data["region_slugs"][(document.getElementById("mirrored_page_region") as HTMLInputElement)?.value];
```
This selector does not seem to work for me, I have selected a mirrored page from augsburg, but it tries to fetch the content from nurnberg...

timoludwig replied:
This selector does not seem to work for me, I have selected a mirrored page from augsburg, but it tries to fetch the content from nurnberg...

membralala commented:
Hey Timo, thank you for your feedback! 

> Hmm, the content of the mirrored page is available in the template context via
> {% get_translation page.mirrored_page language.slug as mirrored_page_translation %}
> so maybe this can be inserted into the popup during template rendering, so it does not have to be inserted via JS?

But this approach would mean, that the preview for the selected Live content is not available until the form was saved, am I right? I thought a while about this and I felt this would be a bit inconsistent, as the preview shows title and tinymce content in the current state of editing, even if they weren't saved.  
Or is there something that I'm missing with your suggestion?

membralala commented:
@timoludwig, I slightly changed the way form values are associated with the corresponding mirrored page content and added handling for some edge cases, which should hopefully fix the issues you mentioned.
I also added text alignment for languages, which are read from right to left.

timoludwig reviewed (commented):
Thanks a lot! :+1:
Everything seems to work now, very cool! :rocket:

I still see a few things that could be improved, have a look at my comments below for details.

Apart from that, could you start resolving all the conflicts and try to fix all CI errors etc (and add a note to the changelog :wink:)?

Thanks!

membralala commented on integreat_cms/cms/views/pages/page_actions.py:
```diff
@@ -157,32 +158,61 @@ def view_page(request, page_id, region_slug, language_slug):
-            "mirrored_translation": mirrored_translation,
+    return JsonResponse(
+        data={
+            "title": page_translation.title if page_translation else "",
```
Is it a possibility, that there is no page content, but still embedded live content?

timoludwig replied:
Yes, but this is only shown in the app when the page translation object actually exists. When the page translation exists and has an empty content, the live content is embedded. But when the page translation does not exist at all, the page is just invisible in the app for this language...

membralala commented:
It seems that tinymce api for react does not quite work after rebasing. ```tinymce.activeEditor``` as well as ```tinymce.get("tinymce")``` returns always null. I'm now reading the tinymce content by accessing the iframes contentDocument. Does this work for you now as well?

timoludwig commented:
> It seems that tinymce api for react does not quite work after rebasing. `tinymce.activeEditor` as well as `tinymce.get("tinymce")` returns always null. I'm now reading the tinymce content by accessing the iframes contentDocument. Does this work for you now as well?

No, I don't think the API of tinymce changed, since I don't see a according entry in the [changelog](https://www.tiny.cloud/docs/changelog/) and we use it in [other places](https://github.com/digitalfabrik/integreat-cms/blob/d52ca0de5ac5c543df74d464ea64e7f29f4a8dce/integreat_cms/static/src/js/forms/tinymce-init.ts#L26) without problems...
I rather think this has something to do with the scope, maybe you should add a small helper function
```
export function getContent(): string {
  return tinymce.activeEditor.getContent();
}
```
to [integreat_cms/static/src/js/forms/tinymce-init.ts](https://github.com/digitalfabrik/integreat-cms/blob/develop/integreat_cms/static/src/js/forms/tinymce-init.ts) which you can then import in `page-preview.ts`:
```
import { getContent } from "../forms/tinymce-init";

...

const pageContent = getContent();
```
I think this would be preferable to the iframe solution.

timoludwig commented:
> Would it make sense to allow previewing empty pages? Once #1371 gets closed editors might want to be able view the list of links to alternatives.

I'm not sure, maybe in the future but for this PR I don't think we should bother :sweat_smile:

membralala commented:
@timoludwig  thanks for your feedback, I changed it as you suggested. I still wonder, why the scoping was an issue there (and especially why it worked fine before rebasing), so if you'd have any further explanation, I'd be  happy to learn something :D

timoludwig commented:
> @timoludwig thanks for your feedback, I changed it as you suggested. I still wonder, why the scoping was an issue there (and especially why it worked fine before rebasing), so if you'd have any further explanation, I'd be happy to learn something :D

I'm sorry, I don't completely understand that either :sweat_smile: 
Maybe, because your script `page-preview.ts` is contained in `index.ts` but all the tinymce functionality is contained in `editor.ts`?

I noticed another problem which might be connected with this: The current state of your PR re-introduces the bug #1329 where the tinymce content is reset after submitting. If I remember correctly, this is caused by tinymce initializing twice... So I'm afraid you need yet another small round of refactoring where you split your file into two files - one for the list preview and the shared functions, and one for the form preview. And the latter needs to be included in [integreat_cms/static/src/editor.ts](https://github.com/digitalfabrik/integreat-cms/blob/develop/integreat_cms/static/src/editor.ts)...

timoludwig commented:
@membralala I now split the preview file into two parts and added a `MirroredPageFieldWidget` to calculate the preview urls in python, so we don't have to hardcode them in JS.

ulliholtgrave reviewed (approved):
Thanks, that's a very nice feature! :100:
